core/GlobalState.h

```diff
@@ -320,6 +321,10 @@ class GlobalState final {
 
     bool shouldReportErrorOn(Loc loc, ErrorClass what) const;
 
+    void clearErrorCacheForFile(core::FileRef fref,
+                                std::function<bool(const std::unique_ptr<core::ErrorQueueMessage> &)> predicate);
```

Comment:
>A better alternative is simply to dereference the pointer and pass that in: const T &.

👍 

>If I'm reading the code right, it shouldn't actually be the case that you need to ask "is the code greater than some number"—you can achieve everything in terms of less than comparisons: for namer, < 5000. For resolver, < 6000 and don't worry if a new 40XX error has snuck in—flush it anyways (should never happen, and you can try to enforce it somewhere if you want). For typecheck, pass UINT16_MAX.

I'm fine with passing two int boundaries. Or making this method private and creating three public methods like `clearNamerErrors`, `clearResolverErrors` and `clearInferErrors`. 

Here is an example why two boundaries are required here:
Imagine you report an error in namer in file `a.rb`. It's flushed after namer finishes and user can see it in the editor. The error goes into cache. Then Sorbet runs resolver and finds another error in file `a.rb`. Now we need to clear the old resolver errors from cache. If we just do `< 6000`, we would also clear the namer error in the same file, and it would disappear from user's editor. If we specify a range of errors to clear, namer error would stay in cache and would be re-published after resolver

Reply:
ok, let's leave it as is for now then, but i still think that passing in an Error or ErrorClass would be better

it might be the case that after some of the changes in service of not caching errors on GlobalState that doing this gets easier.

For example, I had imagined a pair of methods like "flushAndCache" and "flushAndDrop," where the first would flush the error to the LSP client, but keep it in a cache, and subsequent calls to `flushAndCache` would flush both the cached errors *and* any new errors. Then `flushAndDrop` would flush any cached errors + any new errors, and then clear the cache for that file.

so it would look like

```
runNamer()
flushAndCache()
runResolver()
flushAndCache()
runTypecheck()
flushAndDrop()
```

Reply:
Sounds good! My only correction would be to replace last `flushAndDrop` with `flushAndCache`. 
I've tried clearing all the caches after final flush, and it misses errors in case of slow path cancelation. 

My mental model for cache invalidation strategy here is "before flushing {stage} errors, remove {stage} errors from the cache". In that case we would never report stale errors from cache

Reply:
the errors need to get dropped at some point, right? we don't retain them indefinitely I hope? so in that case when do they get dropped?

Reply:
I guess you're saying that they get dropped just before the next time we would populate the cache, e.g. on a second run?

Reply:
We never fully drop the cache, because of the slow path cancelation issue. But we clean irrelevant errors right before reporting after each stage. 
The cache would have something in it almost all of the time, but it would by the number of errors reported in the last slow path